### ISSUE TITLE
lack of suffix ! handling in template.js causing errors 

### ISSUE THREAD

**Issue by mindon** [Polymer/lit-html#613]
> The syntax is suffix !: **identifier!** removes null and undefined from the type of identifier  (TypeScript)

without this part of logic in the generated  template.js, it's causing errors...

line 88 in **template.ts**
`const name = lastAttributeNameRegex.exec(stringForPart)![2];`

v.s.

line 74 in **template.js**
`const name = lastAttributeNameRegex.exec(stringForPart)[2];`

and

line 102 in **template.ts**
`const attributeValue = node.getAttribute(attributeLookupName)!;

const strings = attributeValue.split(markerRegex);`

v.s.

line 87 in **template.js**
`const attributeValue = node.getAttribute(attributeLookupName);

const strings = attributeValue.split(markerRegex);`

**Comment by justinfagnani** [comment]
The `!` suffix is TypeScript only, it won't appear in the JS output.

What is the problem you're seeing? Do you have example input and a stack trace?

**Comment by mindon** [comment]
when `lastAttributeNameRegex.exec(stringForPart)` or `node.getAttribute(attributeLookupName)` is null ...

**Comment by justinfagnani** [comment]
That usually is indicator of a problem with the template. Can you post more details?

**Comment by mindon** [comment]
the case is with my **test/unit/mif-table.html** on https://github.com/mindon/mif-polymer

It may relative with renderring a table, and there's another issue on its testing:

through **test/unit/** the td content rendering empty, and through **test/unit/mif-table.html** the td content rendering ok

**Comment by justinfagnani** [comment]
I don't see any lit-html in test/unit/mif-table.html

Do you have a reproduction of the template that's breaking? I don't think we should default to an empty string, as this error indicates a real problem with a template. We should improve the error message.

**Comment by justinfagnani** [comment]
I'm sorry, this isn't helping much. I see lit-html in the stack, but I don't see a lit-html template in the source you pointed me to. I can't really dig around through a project I'm unfamiliar with. Can you make a small reproduction or at least point me to the lit-html template that's causing the problem?

**Comment by mindon** [comment]
here's a simple case testing code reproducing the bug

[lit-html-bug.zip](https://github.com/Polymer/lit-html/files/2602537/lit-html-bug.zip)

**Comment by jridgewell** [comment]
Reducing this down to just the lit-html parts, there doesn't seem to be any issue. https://output.jsbin.com/muwulod/quiet

**Comment by pitus** [comment]
I just encountered a similar problem when the template contains something like this:
`<my-elem .tags=${tags}></my-elem>` even if I use `.tags="${tags}"`
and the `my-elem` has a `tags` property of type `Object`.
If `tags` value is null (or if I simply use `.tags="${null}"`), then the `attributeValue.split(markerRegex)` throws an exception in `template.js:87` as it's trying to call split on null.

This is, however very odd and took me a while to partially figure out...
The `<my-element>` was within a `<td>` of a table and next row had a missing `</td>` that should have followed another custom control referencing the same `.tags={tags}` property. Once I've added the missing `</td>` the error would no longer be triggered.
The weird thing is that this doesn't seem to happen all the time, after some editing and reverting back to the original problem, I got no error anymore, then it came back again after some further editing, yet the template looked exactly the same, missing `</td>` ... so I'm unable to come up with a minimum test-case to demonstrate this problem yet...
